### ISSUE TITLE
cannot gather or track all outputs from multi gpu settings 

### ISSUE THREAD

**Issue by bocopilot** [Lightning-AI/lightning#18000]
### Bug description

Hi I'm facing an issue in gathering all the losses and predictions in multi gpu scenario. I'm using pytorch lightning 2.0.4 and deepspeed, distributed strategy - deepspeed_stage_2. 

I'm adding my skeleton code here for reference. 

```
    def __init__(self):
        self.batch_train_preds = []
        self.batch_train_losses = []


    def  training_step(self, batch, batch_idx):
        input_ids = batch['input_ids']
        attention_mask = batch['attention_mask']

        # Model Step
        outputs = self.model(input_ids=input_ids, attention_mask=attention_mask, labels=train_labels)

        train_preds = torch.argmax(outputs.logits, dim=-1)

        return {'loss': outputs[0],
                'train_preds': train_preds}

    def on_train_batch_end(self, outputs, batch, batch_idx):
        # aggregate metrics or outputs at batch level
        train_batch_loss = outputs["loss"].mean()
        train_batch_preds = torch.cat(outputs["train_preds"])

        self.batch_train_preds.append(train_batch_preds)
        self.batch_train_losses.append(train_batch_loss.item())

        return {'train_batch_loss': train_batch_loss,
                'train_batch_preds': train_batch_preds
                }

    def on_train_epoch_end(self) -> None:
        # Aggregate epoch level training metrics

        epoch_train_preds = torch.cat(self.batch_train_preds)
        epoch_train_loss = np.mean(self.batch_train_losses)

        self.logger.log_metrics({"epoch_train_loss": epoch_train_loss})

```

In the above code block, I'm trying to combine all the predictions into a single tensor at the end of the epoch by tracking each batch in a global list (defined at init). but in multi gpu training, I faced an error with concatination as each gpu is treating the batch in it's own device and I couldn't combine the results in a single global list. 

Question: 

What should I be doing in on_train_batch_end or on_train_epoch_end or in training_step in order to combine the results across all the gpus into a list created in my init because I want to calculate some additional metrics(precision, recall etc) during ON_*_EPOCH_END() function in my train, validation, test(validation and test is similar to my training functions above). 

### What version are you seeing the problem on?

v2.0

### Error messages and logs

```
# Error messages and logs here please
```

### Environment

<details>
  <summary>Current environment</summary>

```
#- Lightning Component (e.g. Trainer, LightningModule, LightningApp, LightningWork, LightningFlow):
#- PyTorch Lightning Version (e.g., 1.5.0):
#- Lightning App Version (e.g., 0.5.2):
#- PyTorch Version (e.g., 2.0):
#- Python version (e.g., 3.9):
#- OS (e.g., Linux):
#- CUDA/cuDNN version:
#- GPU models and configuration:
#- How you installed Lightning(`conda`, `pip`, source):
#- Running environment of LightningApp (e.g. local, cloud):
```

</details>

**Comment by leng-yue** [comment]
See https://pytorch.org/docs/stable/distributed.html

**Comment by bocopilot** [comment]
@leng-yue I need some help in understanding all_gather, I tried using it and it's actually combining the results across the devices, however, this is happening in every device which is fine. How can I only use one device's output to do some operation(like storing in a list etc in my case) in all devices after using all_gather. 

Also need some guidance on when should I ideally use it i.e at step level or batch end level or epoch end level? I'm assuming calling at batch level like i did would do the job

**Comment by leng-yue** [comment]
You can use gather instead of all_gather if you only want the result to be on one device. By the way, in your above use case, you only need to use self.logger with on_epoch=True. It will automatically calculate the avg loss and log for you.

**Comment by bocopilot** [comment]
I dont want to use self.logger, I want to calculate the metric without self.logger, Since I'm going to log this on mlflow

**Comment by leng-yue** [comment]
The best practice is probably implementing a mlflow logger instead of ignoring the existing logging system...

**Comment by bocopilot** [comment]
@leng-yue I want to setup a mlflow logger but I couldnt get my head around on if I can achieve epoch level logging just like self.log() but using mlflow logger

**Comment by awaelchli** [comment]
@bocopilot See my answer here in the forum, showing how to use `self.all_gather` to combine loss or predictions from multiple processes/GPUs into one result you can log:
https://lightning.ai/forums/t/combining-loss-predictions-in-multi-gpus/3185/2?u=awaelchli

Please let me know if this solves your issue.
